File: backend/simulators/financialFraudSimulator.js
// Financial Fraud/Anomaly Investigation simulator — the actual cause behind
// an unusual transaction pattern, not just a flagged correlation. Includes
// a genuine "not fraud" class — real fraud detection has to distinguish
// fraud from legitimate-but-unusual behavior, not just classify among
// fraud types.

const CAUSES = ['stolen_card_testing', 'account_takeover', 'synthetic_identity', 'merchant_collusion', 'legitimate_unusual_purchase'];

const SIGNAL_PROFILES = {
  stolen_card_testing:         { rapid_small_transactions: 0.9, velocity_spike: 0.85, new_shipping_address: 0.2, device_fingerprint_mismatch: 0.3, high_risk_merchant_category: 0.4, geo_mismatch: 0.3 },
  account_takeover:            { device_fingerprint_mismatch: 0.85, geo_mismatch: 0.7, new_shipping_address: 0.6, velocity_spike: 0.4, rapid_small_transactions: 0.15, high_risk_merchant_category: 0.2 },
  synthetic_identity:          { new_shipping_address: 0.7, high_risk_merchant_category: 0.3, velocity_spike: 0.25, device_fingerprint_mismatch: 0.2, rapid_small_transactions: 0.1, geo_mismatch: 0.15 },
  merchant_collusion:          { high_risk_merchant_category: 0.8, rapid_small_transactions: 0.3, velocity_spike: 0.3, new_shipping_address: 0.05, device_fingerprint_mismatch: 0.1, geo_mismatch: 0.1 },
  legitimate_unusual_purchase: { new_shipping_address: 0.25, geo_mismatch: 0.2, velocity_spike: 0.1, high_risk_merchant_category: 0.1, rapid_small_transactions: 0.03, device_fingerprint_mismatch: 0.05 },
};
const ALL_SIGNALS = ['rapid_small_transactions', 'new_shipping_address', 'device_fingerprint_mismatch', 'velocity_spike', 'high_risk_merchant_category', 'geo_mismatch'];

function sampleIntervention() { return { cause: CAUSES[Math.floor(Math.random() * CAUSES.length)] }; }
function generateObservations(intervention) {
  const profile = SIGNAL_PROFILES[intervention.cause];
  const observations = {};
  for (const s of ALL_SIGNALS) observations[s] = Math.random() < (profile[s] ?? 0.02);
  return observations;
}
function verifyProposedStructure(proposed, intervention) {
  if (!proposed || typeof proposed !== 'object' || !('cause' in proposed)) return 0.0;
  if (proposed.cause === intervention.cause) return 1.0;
  const trueProfile = SIGNAL_PROFILES[intervention.cause] || {};
  const proposedProfile = SIGNAL_PROFILES[proposed.cause] || {};
  let overlap = 0, total = 0;
  for (const s of ALL_SIGNALS) { overlap += Math.min(trueProfile[s] || 0, proposedProfile[s] || 0); total += trueProfile[s] || 0; }
  return total > 0 ? Math.round((0.3 * (overlap / total)) * 1000) / 1000 : 0.0;
}
function proposeStructureHeuristic(observations) {
  const present = ALL_SIGNALS.filter(s => observations[s]);
  if (present.length === 0) return { cause: 'legitimate_unusual_purchase' };   // no red flags -> assume legitimate, matching real-world fraud triage
  let best = CAUSES[0], bestScore = -1;
  for (const c of CAUSES) {
    const score = present.reduce((sum, s) => sum + (SIGNAL_PROFILES[c][s] || 0), 0);
    if (score > bestScore) { bestScore = score; best = c; }
  }
  return { cause: best };
}
function generateScenarios(count) {
  const out = [];
  for (let i = 0; i < count; i++) { const intervention = sampleIntervention(); out.push({ hidden_intervention: intervention, observations: generateObservations(intervention) }); }
  return out;
}
function runSingleScenario(observations, hiddenIntervention) {
  const proposed = proposeStructureHeuristic(observations);
  return { proposed, reward: verifyProposedStructure(proposed, hiddenIntervention) };
}
module.exports = { generateScenarios, runSingleScenario, sampleIntervention, generateObservations, verifyProposedStructure, proposeStructureHeuristic, CAUSES, ALL_SIGNALS, SIGNAL_PROFILES };

File: training/simulators/financial_fraud_simulator.py
"""Financial Fraud/Anomaly Investigation simulator — Python reference
version, mirrors backend/simulators/financialFraudSimulator.js exactly.
Includes a genuine 'not fraud' class — see the JS version's comment."""
import random

CAUSES = ["stolen_card_testing", "account_takeover", "synthetic_identity", "merchant_collusion", "legitimate_unusual_purchase"]
SIGNAL_PROFILES = {
    "stolen_card_testing":         {"rapid_small_transactions": 0.9, "velocity_spike": 0.85, "new_shipping_address": 0.2, "device_fingerprint_mismatch": 0.3, "high_risk_merchant_category": 0.4, "geo_mismatch": 0.3},
    "account_takeover":            {"device_fingerprint_mismatch": 0.85, "geo_mismatch": 0.7, "new_shipping_address": 0.6, "velocity_spike": 0.4, "rapid_small_transactions": 0.15, "high_risk_merchant_category": 0.2},
    "synthetic_identity":          {"new_shipping_address": 0.7, "high_risk_merchant_category": 0.3, "velocity_spike": 0.25, "device_fingerprint_mismatch": 0.2, "rapid_small_transactions": 0.1, "geo_mismatch": 0.15},
    "merchant_collusion":          {"high_risk_merchant_category": 0.8, "rapid_small_transactions": 0.3, "velocity_spike": 0.3, "new_shipping_address": 0.05, "device_fingerprint_mismatch": 0.1, "geo_mismatch": 0.1},
    "legitimate_unusual_purchase": {"new_shipping_address": 0.25, "geo_mismatch": 0.2, "velocity_spike": 0.1, "high_risk_merchant_category": 0.1, "rapid_small_transactions": 0.03, "device_fingerprint_mismatch": 0.05},
}
ALL_SIGNALS = ["rapid_small_transactions", "new_shipping_address", "device_fingerprint_mismatch", "velocity_spike", "high_risk_merchant_category", "geo_mismatch"]

def sample_intervention(rng): return {"cause": rng.choice(CAUSES)}
def generate_observations(intervention, rng):
    profile = SIGNAL_PROFILES[intervention["cause"]]
    return {s: rng.random() < profile.get(s, 0.02) for s in ALL_SIGNALS}
def verify_proposed_structure(proposed, intervention, rng=None, n_trials=20):
    if not isinstance(proposed, dict) or "cause" not in proposed: return 0.0
    if proposed["cause"] == intervention["cause"]: return 1.0
    true_p, prop_p = SIGNAL_PROFILES[intervention["cause"]], SIGNAL_PROFILES.get(proposed["cause"], {})
    overlap = sum(min(true_p.get(s, 0), prop_p.get(s, 0)) for s in ALL_SIGNALS)
    total = sum(true_p.values())
    return round(0.3 * (overlap / total), 3) if total > 0 else 0.0

if __name__ == "__main__":
    import sys, json
    if "--generate" in sys.argv:
        n = int(sys.argv[sys.argv.index("--generate") + 1]); rng = random.Random(); out = []
        for _ in range(n):
            iv = sample_intervention(rng); out.append({"hidden_intervention": iv, "observations": generate_observations(iv, rng)})
        print(json.dumps(out)); sys.exit(0)
    rng = random.Random(42); correct, wrong_rewards = 0, []
    for _ in range(200):
        iv = sample_intervention(rng); guess = {"cause": "stolen_card_testing"}
        r = verify_proposed_structure(guess, iv)
        if r == 1.0: correct += 1
        elif iv["cause"] != "stolen_card_testing": wrong_rewards.append(r)
    print(f"Self-test: {correct}/200 correct, wrong-guess reward range {min(wrong_rewards):.3f}-{max(wrong_rewards):.3f}")
    assert 0 < correct < 200
    print("Simulator self-test passed ✓")
